### PR TITLE
Add pool op

### PR DESCRIPTION
**Pull request by chengduoZH**
Add pool op.

- [x] Add pooling2d max and average (gpu and cpu implementations)
- [x] Add pooling3d max and average (gpu and cpu implementations)
- [x] Add unit test pooling2d and pooling3d

fixed https://github.com/PaddlePaddle/Paddle/issues/4231
fixed https://github.com/PaddlePaddle/Paddle/issues/3692
fixed https://github.com/PaddlePaddle/Paddle/issues/4165
fixed https://github.com/PaddlePaddle/Paddle/issues/4164

### CODE REVIEW

File: paddle/operators/pool_op.cc [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+class PoolOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("Input"),
+                            "Input(Input) of Pooling should not be null.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("Output"),
```
Comment (NHZlX) [inline paddle/operators/pool_op.cc]: 这里应该是 ctx.OutputVar("Output") 吧

Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: 是的，谢谢！

File: paddle/operators/pool_op.cc [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        "number of channels, H and W is the height and width of image.");
+    AddOutput("Out",
+              "The output tensor of pooling operator."
+              "The format of output tensor is also NCHW.");
+
+    AddAttr<std::string>("pooling_type",
+                         "pooling_type of pooling operator.['max' or 'ave']");
+    AddAttr<std::vector<int>>(
+        "ksize", "pooling size(height, width) of pooling operator.");
+    AddAttr<int>("global_pooling", "whether to use the global_pooling.[0 or 1]")
```
Comment (chengduoZH) [inline paddle/operators/pool_op.cc]: "pooling_type" and "ksize"  don't set default values. Should it be judged in InferShape?@hedaoyuan

File: paddle/operators/math/pooling.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+class maxPool {
+ public:
+  DEVICE inline T initial() { return static_cast<T>(-FLT_MAX); }
+  DEVICE inline void process(T& y, const T& x) { y = y > x ? y : x; }
+  DEVICE inline void finalize(T& y, const T& poo_size) {}
+  DEVICE inline void gradProcess(const T& x, const T& y, const T& dy, T& dx,
+                                 T scale, bool& stop) {
+    if (stop == false) {
+      stop = (x == y);
+      dx += dy * (x == y);
```
Comment (hedaoyuan) [inline paddle/operators/math/pooling.h]: 可以改成：`if (x == y) dx += dy;`
另外，这个`if (stop == false)`并不是很有必要，直接修改成`if (x == y) `计算量是一样的。

File: paddle/operators/math/pooling.h [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    }
+  }
+};
+
+template <class T>
+class avePool {
```
Comment (hedaoyuan) [inline paddle/operators/math/pooling.h]: 感觉forward/backward计算没必要写到一个类里面，可以直接写两个类avgPool和avgPoolGrad。

File: paddle/operators/math/pooling.cu [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+        PoolProcess,
+        T><<<grid, threads, 0,
+             reinterpret_cast<const platform::CUDADeviceContext&>(context)
+                 .stream()>>>(nthreads, input_data, output_data, input_channels,
+                              input_height, input_width, output_height,
+                              output_width, ksize_height, ksize_width,
+                              stride_height, stride_width, padding_height,
+                              padding_width, pool_process);
+
+    // CHECK_SYNC("Pool2dForwardKernel failed");
```
[the remark this answers is not in the document]
Comment (chengduoZH) [reply paddle/operators/math/pooling.cu]: Done

File: paddle/operators/math/pooling.cu [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+        PoolProcess,
+        T><<<grid, threads, 0,
+             reinterpret_cast<const platform::CUDADeviceContext&>(context)
+                 .stream()>>>(
+        nthreads, input_data, output_data, output_grad_data, input_grad_data,
+        input_channels, input_height, input_width, output_height, output_width,
+        ksize_height, ksize_width, stride_height, stride_width, padding_height,
+        padding_width, pool_process);
+
+    // CHECK_SYNC("KernelPool2dBackward failed");
```
[the remark this answers is not in the document]
Comment (chengduoZH) [reply paddle/operators/math/pooling.cu]: Done

File: paddle/operators/math/pooling.cu [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+        T><<<grid, threads, 0,
+             reinterpret_cast<const platform::CUDADeviceContext&>(context)
+                 .stream()>>>(
+        nthreads, input_data, output_data, input_channels, input_depth,
+        input_height, input_width, output_depth, output_height, output_width,
+        ksize_depth, ksize_height, ksize_width, stride_depth, stride_height,
+        stride_width, padding_depth, padding_height, padding_width,
+        pool_process);
+
+    // CHECK_SYNC("Pool2dForwardKernel failed");
```
[the remark this answers is not in the document]
Comment (chengduoZH) [reply paddle/operators/math/pooling.cu]: Done

File: paddle/operators/math/pooling.cu [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+        T><<<grid, threads, 0,
+             reinterpret_cast<const platform::CUDADeviceContext&>(context)
+                 .stream()>>>(
+        nthreads, input_data, output_data, output_grad_data, input_grad_data,
+        input_channels, input_depth, input_height, input_width, output_depth,
+        output_height, output_width, ksize_depth, ksize_height, ksize_width,
+        stride_depth, stride_height, stride_width, padding_depth,
+        padding_height, padding_width, pool_process);
+
+    // CHECK_SYNC("KernelPool2dBackward failed");
```
[the remark this answers is not in the document]
Comment (chengduoZH) [reply paddle/operators/math/pooling.cu]: Done

File: paddle/operators/math/pooling.h (hunk 3 above)
Comment (chengduoZH) [reply paddle/operators/math/pooling.h]: fixed

File: paddle/operators/math/pooling.h (hunk 4 above)
Comment (chengduoZH) [reply paddle/operators/math/pooling.h]: Done

File: paddle/operators/math/pooling.h [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+namespace paddle {
+namespace operators {
+namespace math {
+
+//////////////////////
+#define FLT_MAX __FLT_MAX__
+/////////////////////
+
+namespace pool {
```
Comment (qingqing01) [inline paddle/operators/math/pooling.h]: 如果要加`namespace pool`,需要新建`paddle/operators/math/pool` 路径。

File: paddle/operators/math/pooling.cc [hunk 10]
```diff
@@ -0,0 +1,470 @@
...
+template class MaxPool2dBackwardFunctor<platform::CPUPlace, float>;
+// template class MaxPool2dBackwardFunctor<platform::CPUPlace, double>;
+
+template class Pool2dForwardFunctor<
+    platform::CPUPlace, paddle::operators::math::pool::maxPool<float>, float>;
+template class Pool2dForwardFunctor<
+    platform::CPUPlace, paddle::operators::math::pool::avgPool<float>, float>;
+template class Pool2dBackwardFunctor<
+    platform::CPUPlace, paddle::operators::math::pool::maxPoolGrad<float>,
+    float>;
```
Comment (qingqing01) [inline paddle/operators/math/pooling.cc]: 这个和line 198 `MaxPool2dBackwardFunctor`有啥区别？

File: paddle/operators/pool_op.cc [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/operators/pool_op.h"
+
+namespace paddle {
+namespace operators {
+
+int outputSize_pool(int input_size, int filter_size, int padding, int stride) {
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: Function_Names:  "Camel Case"

https://google.github.io/styleguide/cppguide.html#Function_Names

File: paddle/operators/pool_op.cc [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("X"),
+                            "X(Input) of Pooling should not be null.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.OutputVar("Out"),
+                            "Out(Output) of Pooling should not be null.");
+
+    auto in_X = ctx.Input<Tensor>("X");
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: in_X -> in_x

File: paddle/operators/pool_op.cc [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+    auto out = ctx.Output<Tensor>("Out");
+    int global_pooling = Attr<int>("globalPooling");
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: 可以使用bool type了。

File: paddle/operators/pool_op.cc [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+      PADDLE_ENFORCE_EQ(strides.size(), 2,
+                        "Pool2DOp strides size should be 2 elements.");
+      PADDLE_ENFORCE_EQ(paddings.size(), 2,
+                        "Pool2DOp paddings size should be 2 elements");
+    } else {
+      PADDLE_ENFORCE_EQ(strides.size(), 3,
+                        "Pool3DOp strides should be 3 elements.");
+      PADDLE_ENFORCE_EQ(paddings.size(), 3,
+                        "Pool3DOp paddings should be 3 elements.");
+    }
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: line 55 - line 65的11行写成下面3行即可：
```cpp
PADDLE_ENFORCE(ksize.size() == 2 || ksize.size() == 3);
PADDLE_ENFORCE_EQ(ksize.size(), strides.size());
PADDLE_ENFORCE_EQ(ksize.size(), paddings.size());
```

File: paddle/operators/pool_op.cc [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+    out->Resize(framework::make_ddim(output_shape));
+  }
+};
+
+class PoolOpGrad : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: 和forward一样，加PADDLE_ENFORCE_NOT_NULL检查。

File: paddle/operators/pool_op.cc [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+    auto in = ctx.Input<Tensor>("X");
+    auto d_in = ctx.Output<Tensor>(framework::GradVarName("X"));
+    if (d_in) d_in->Resize(in->dims());
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: backward只有一个输出的，可以不用判断，直接写成：

```cpp
d_in->Resize(in->dims());
```

File: paddle/operators/pool_op.cc [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+        "whether to use the globalPooling."
+        "int constant equal to 0 or 1"
+        "default 0"
+        "If globalPooling = 1, ksize is ignored and need not be specified.")
+        .SetDefault(0);
+    AddAttr<std::vector<int>>("strides",
+                              "strides(height, width) of pooling operator."
+                              "default {1,1}")
+        .SetDefault({1, 1})
+        .AddCustomChecker(GreaterThanChecker_pool({0, 0}));
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: `GreaterThanChecker_pool({1, 1}`

File: paddle/operators/pool_op.cc [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+        .SetDefault({0, 0})
+        .AddCustomChecker(EqualGreaterThanChecker_pool({0, 0}));
+    AddComment(R"DOC(
+The pooling2d operation calculates the output based on
+the input, poolingType and ksize, strides, paddings parameters.
+)DOC");
+  }
+
+ private:
+  struct GreaterThanChecker_pool {
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: 命名规则：https://google.github.io/styleguide/cppguide.html#Type_Names

 另外可以直接给`paddle/framework/attribute.h`里加吧。

File: paddle/operators/pool_op.cc [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+class Pool3dOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  Pool3dOpMaker(framework::OpProto *proto, framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X",
+             "The input tensor of pooling operator. "
+             "The format of input tensor is NCDHW. Where N is batch size, C is "
+             "the "
+             "number of channels, D, H and W is the depth, height and width of "
+             "image.");
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: image -> feature.

File: paddle/operators/pool_op.cc [hunk 20]
```diff
@@ -0,0 +1,255 @@
...
+              "The output tensor of pooling operator."
+              "The format of output tensor is also NCHW.");
+
+    AddAttr<std::string>("poolingType",
+                         "poolingType of pooling operator."
+                         "str constant equal to 'max' or 'avg'");
+    AddAttr<std::vector<int>>(
+        "ksize", "pooling size(height, width) of pooling operator.")
+        .AddCustomChecker(GreaterThanChecker_pool({0, 0}));
+    AddAttr<int>(
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: use bool type.

File: paddle/operators/pool_op.cc (hunk 11 above)
Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: Done

File: paddle/operators/pool_op.cc (hunk 12 above)
Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: Done

File: paddle/operators/pool_op.cc (hunk 14 above)
Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: Done

File: paddle/operators/pool_op.cc (hunk 15 above)
Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: Done

File: paddle/operators/pool_op.cc (hunk 16 above)
Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: Done

File: paddle/operators/pool_op.cc (hunk 17 above)
Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: "strides"是可以等于1的

File: paddle/operators/pool_op.cc (hunk 19 above)
Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: Done

File: paddle/operators/math/pooling.h (hunk 9 above)
Comment (chengduoZH) [reply paddle/operators/math/pooling.h]: Done

File: paddle/operators/math/pool_test_maxPool2d.cc [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+    maxpool2d_backward(context, input, input_grad, output, output_grad, ksize,
+                       strides, paddings);
+    PADDLE_ENFORCE(
+        cudaStreamSynchronize(0),
+        "cudaStreamSynchronize failed in maxpool2d_backward CopyFrom");
+  }
+  finish = clock();
+  totaltime = (double)(finish - start) / CLOCKS_PER_SEC;
+  totaltime /= times;
+  std::cout << "\nMaxPool3dBackwardFunctor: " << totaltime << "s" << std::endl;
```
[the remark this answers is not in the document]
Comment (chengduoZH) [reply paddle/operators/math/pool_test_maxPool2d.cc]: Done

File: paddle/operators/math/pooling.cc [hunk 22]
```diff
@@ -0,0 +1,470 @@
...
+limitations under the License. */
+
+#include "paddle/operators/math/pooling.h"
+
+namespace paddle {
+namespace operators {
+namespace math {
+
+template <typename PoolProcess, typename T>
+class Pool2dForwardFunctor<platform::CPUPlace, PoolProcess, T> {
```
[the remark this answers is not in the document]
Comment (chengduoZH) [reply paddle/operators/math/pooling.cc]: rename `Pool2dForwardFunctor ` -> `Pool2dFunctor `, `Pool2dBackwardFunctor ` -> `Pool2dGradFunctor `

File: python/paddle/v2/framework/tests/test_pool3d_op.py [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+        self.op_type = "pool3d"
+        self.pool_type = "max"
+        self.pool3D_forward_naive = max_pool3D_forward_naive
+        self.shape = [2, 3, 5, 5, 5]
+        self.ksize = [3, 3, 3]
+        self.strides = [1, 1, 1]
+        self.paddings = [0, 0, 0]
+
+
+class TestCase3(TestPool3d_Op):
```
Comment (hedaoyuan) [inline python/paddle/v2/framework/tests/test_pool3d_op.py]: TestCase4?

File: python/paddle/v2/framework/tests/test_pool2d_op.py [hunk 24]
```diff
@@ -0,0 +1,144 @@
...
+        self.op_type = "pool2d"
+        self.pool_type = "max"
+        self.pool2D_forward_naive = max_pool2D_forward_naive
+        self.shape = [2, 3, 5, 5]
+        self.ksize = [3, 3]
+        self.strides = [1, 1]
+        self.paddings = [0, 0]
+
+
+class TestCase3(TestPool2d_Op):
```
Comment (hedaoyuan) [inline python/paddle/v2/framework/tests/test_pool2d_op.py]: TestCase3 repeated definition

Comment (chengduoZH) [reply python/paddle/v2/framework/tests/test_pool2d_op.py]: Done

File: python/paddle/v2/framework/tests/test_pool3d_op.py (hunk 23 above)
Comment (chengduoZH) [reply python/paddle/v2/framework/tests/test_pool3d_op.py]: Done

**Pull request merged**